### ISSUE TITLE
Compliance Tagging

### ISSUE THREAD

**Issue by brian-ruf** [usnistgov/OSCAL#658]
# User Story:

As an OSCAL Framework Compliance Architect, I need a consistent mechanism to identify content in OSCAL files that is intended to meet specific framework requirements. 

A previous FedRAMP approach used prescribed ID values to clearly identify content required by FedRAMP; this approach does not scale to multiple frameworks as each may require different ID values for the same content. This approach may also be incompatible with changes NIST is considering for OSCAL ID fields.

Tools still need a clear/non-ambiguous method of identifying required content for a given framework in cases where core OSCAL content and framework extensions are not sufficient to align OSCAL content with framework requirements.

In many cases, such as with an authorization boundary diagram in an SSP, the core OSCAL syntax is clear and no such mechanism is needed; however, with other information the OSCAL syntax alone is not sufficient. For example, use of the ```location``` assembly in ```metadata``` is used to identify the location of data centers, administrative offices, organizational headquarters, and other possible locations. FedRAMP validation tools must have a consistent mechanism for identifying which location entries are for a primary data center and which are for backup data centers.

## Goals:

- [ ] Provide a clear mechanism for identifying information as valid for individual frameworks.
- [ ] Ensure the mechanism allows multiple frameworks to be represented simultaneously in a single OSCAL file.

## Dependencies:

None. 

## Acceptance Criteria

- [ ] All [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes in this issue have been updated. Changes to the OSCAL website can be made in the docs/content directory of your branch.
- [ ] A Pull Request (PR) is submitted that fully addresses the goals of this User Story. This issue is referenced in the PR.
- [ ] The CI-CD build process runs without any reported errors on the PR. This can be confirmed by reviewing that all checks have passed in the PR.

**Comment by brianrufgsa** [comment]
As discussed verbally in a working session on 22-Apr-2020 between @wendellpiez, @david-waltermire-nist, and @brianrufgsa, this could be achieved by using ```prop``` fields with a designated ```@name``` flag value to "tag" items for compliance purposes. 

Since this approach could be used by multiple frameworks, the ```@ns``` flag would also be used to designate the source of the requirement.

For example, the FedRAMP-required primary data center locations could be tagged as follows:
```
      <location id="loc-1">
         <address>
            <addr-line>1111 Main Street</addr-line>
            <city>Anywhere</city>
            <state>--</state>
            <postal-code>00000-0000</postal-code>
         </address>
         <prop name="conformity" ns="https://fedramp.gov/ns/oscal">primary-data-center</prop>
      </location>
```
A different compliance framework could be free to use their own values for the prop field, and ignore the FedRAMP field as follows:

```
      <location id="loc-1">
         <address>
            <addr-line>1111 Main Street</addr-line>
            <city>Anywhere</city>
            <state>--</state>
            <postal-code>00000-0000</postal-code>
         </address>
         <prop name="conformity" ns="https://fedramp.gov/ns/oscal">primary-data-center</prop>
         <prop name="conformity" ns="http://something.org/ns/oscal">data-center-1</prop>
      </location>
```

**Comment by brianrufgsa** [comment]
If we take the above approach a reserved name must be defined for the ```@name``` flag, which cannot be used for any other purpose in OSCAL, neither as core syntax nor an extension.

Some considerations include:
```
compliance
conformity
requirement
satisfaction
```

I favor "conformity" only because I believe it is unlikely to be needed in any other OSCAL context, yet seems clear that it is tagging data which conforms with a particular framework's requirement.

**Comment by bradh** [comment]
-0 (somewhat opposed). I'm not disagreeing with the potential application, but I really don't like the naming (or perhaps I'm just bike-shedding). 

Conformity / compliance implies matching some form of requirement - so the value should be something about whether it matches the requirement, or how, or something like that. `requirement` and `satisfaction` have equivalent problems.

If I understand what you're trying to do, its more like a clarifying statement or extension, or some linkage to a domain-specific ontology / terminology. That could be used for more than just "I need this to meet a specific requirement". Could it be considered "additional domain specific detail"?

[I've tried to avoid a specific name here, because I'm not pushing a particular solution]

**Comment by wendellpiez** [comment]
> If we take the above approach a reserved name must be defined for the `@name` flag, which cannot be used for any other purpose in OSCAL, neither as core syntax nor an extension.

@brianrufgsa I take it you mean within a FedRAMP processing context? (that is, by a consumer that recognizes and respects such FedRAMP-defined rules?)

It seems to be we have a nice example of a rules cascade here:
- `prop` may have `@name` and `@ns` - **OSCAL**
- when its `@ns` indicates FedRAMP, `location/prop` only permits a given value on `@name` (e.g., `conformity`, `compliance` etc.) - **FedRAMP**
- when its `@ns` indicates FedRAMP, a `location/prop[@name='conformity']` value must be a given value (e.g. `primary-data-center`, `data-center1` etc.) - **FedRAMP**

What would be the impact on other OSCAL users if FedRAMP defines its rules here? As long as that initial clause is respected (scoping the applicability of the rule to FedRAMP) It seems to me that FedRAMP or any user (along with partners and stakeholders) can define these rules how it needs to - make them strict or loose even ("we forbid any `prop` outside our own `@ns`, scrub them out of inputs, and we jealously and enthusiastically validate everything we see") -- without impacting users who choose not to follow them? Maybe I'm missing a wrinkle?

**Comment by brianrufgsa** [comment]
@bradh (I'll confess I had to look up "bike-shedding")
It's fair to not like the naming. I don't love the naming either, which was part of what prompted me to pose the question. 

The goal is to make it easy for tools to find content as it pertains to the requirements of a specific framework. (In this case FedRAMP.) I am open to other solutions and/or names; however, I'm a contractor to FedRAMP and under a deadline, so I have to commit to a solution now and move forward quickly. I'd prefer to move forward with a solution that is less likely to require rework later.

**Comment by brianrufgsa** [comment]
@wendellpiez It was my intention to say that whatever we select for the ```@name``` value be a reserved value in OSCAL. 

Any time a ```prop``` exists with that ```@name```, it should be regarded by OSCAL-compliant tools as a framework-specific compliance tag, and should not appear in OSCAL for any other reason (as the value of an @name flag on a prop field). This is one case where prop @name without @ns is meaningless.

**Comment by wendellpiez** [comment]
If I understand this, I am against it.

Does OSCAL have to extend its list of reserved names for every user constituency that has (reasonable) requirements for identifying and labeling local semantics? If we do this, I am not sure why we are asking people to use `@ns`.

In my mind, `@ns` says "the meaning here was defined by X. See their docs."

Using `prop` without `@ns` means "the meaning here is defined by OSCAL, or by someone else, or not at all -- see your local documentation in reference to OSCAL documentation".

If we want to reserve a vocabulary for broader-than-FedRAMP but not all of OSCAL, we could maybe make a namespace for that? Broadly forbidding values such as `compliance` or any other value from `prop/@name` seems like overkill to me.

However, in passing I note that the next-gen Metaschema will support our declaring and enforcing such constraints if we want to, for example, triggering value tests (on `@name` or a field value or both in combination) based on `@ns` assignment or lack thereof. (At least, that is the idea.)

**Comment by brianrufgsa** [comment]
@wendellpiez fair enough. 
I'm happy to **not** make this an OSCAL feature, and simply make it another FedRAMP extension. Candidly that reduces schedule risk for me. 

I believe; however, that this feature is needed in some form for other frameworks and organizational requirements as well. I'm trying to find a solution that can be consistent across OSCAL and scale to other frameworks.

If I simply make this a FedRAMP extension for now it could limit NIST's ability to address this feature later, since re-work would be significant, and it's possible this approach becomes entrenched as a de facto standard.

I am open to better solutions, but am not hearing any and need to move forward.

**Comment by wendellpiez** [comment]
@brianrufgsa I think I get you now: it is not a matter of FedRAMP, but of helping others Do The Right Thing and reducing wheel-invention.

In my experience these sorts of things (e.g.: developing the common vocabulary to use with the common vocabulary) go somewhat more slowly than we would like mainly because all sides must be heard from, or decisions that were made, don't stick, or have to be unmade.

However, I do think there is a way we could maybe thread the needle on this one. The registry application @david-waltermire-nist has mentioned has potential for this. We could define a namespace for community practice. As you say, working on this together would reduce risk for all participants and concerned parties.

[JATS](https://jats.nlm.nih.gov/) has [JATS4R ](https://jats4r.org/) for this sort of thing -- outside the core standard, which gives flexibility, and user- and community-driven. But with a defined way to push into the standard.

**Comment by david-waltermire-nist** [comment]
This issue was opened prior to the OSCAL 1.0.0 release. The use case of defining locations as data centers has been implemented since OSCAL 1.0.0. See https://pages.nist.gov/OSCAL/reference/latest/system-security-plan/json-reference/#/system-security-plan/metadata/locations.

The rest of this is not sufficiently defined to work on. Closing this issue.
